### DISCUSSION TITLE
Garmin FIT File Format and Laps (Tour Markers)

### DISCUSSION THREAD

**Discussion by GurVJ** [mytourbook/mytourbook#1279] [Q&A]
I am part of the Beta Program for Garmin Forerunner 965 watch and in the last 2 Beta releases of the firmware there is a change in the FIT file format which leads to the fact that there is no more information about laps (or markers) in MyTourBook.
I do not know if it is a firmware bug (I reported the problem but don't have any answer) or if it will be the new way for Garmin to implement FIT file. I also can't guess if it is a bug because with the Garmin Connect software all the informations about Laps (Intervals for Garmin) are available and OK.

Description of the problem :

Tour Markers with MyTourBook (no more delta time and wrong delta distance)
![MyTourBookLaps](https://github.com/mytourbook/mytourbook/assets/64196685/8e9166e1-de7d-4699-bdbf-26372d662f72)

Same data from Garmin Connect software
![GarminConnectLaps](https://github.com/mytourbook/mytourbook/assets/64196685/632abc7d-1ac0-45ac-861a-05c95fddc9f2)

I think the problem comes from the "timestamp" field in "lap" messages which has got now a fixed value in the FIT file :
![lapsV18 12](https://github.com/mytourbook/mytourbook/assets/64196685/46b8159d-7547-46ab-8623-857d6c5c8ee6)

But why can't MyTourBook retreive the data whereas Garmin Connect is able to do it ?

**Comment by FJBDev** [comment]
>But why can't MyTourBook retreive the data whereas Garmin Connect is able to do it ?

Without any file to look at, its going to be hard to answer this question.

**Comment by GurVJ** [comment]
I thought the last screen capture with timestamps would help, but you'll find the original file below.

[2024-01-11-18-31-17.zip](https://github.com/mytourbook/mytourbook/files/14144748/2024-01-11-18-31-17.zip)

**Comment by FJBDev** [comment]
@GurVJ 
Thanks for the file. The screen capture helps but a file is better to dig into the actual possible problems.
But after investigating, it seems that you are right, the "timestamp" field in "lap" messages, which creates this issue in the end.

![image](https://github.com/mytourbook/mytourbook/assets/25235665/6b66b432-3294-4a61-9550-37626d136d32)

I can't explain why it works in Garmin Connect but maybe Garmin generates the lap time from another data ? Not sure...

**Comment by GurVJ** [comment]
After a quick test, it is also working with Strava. So it means that it is not a Garmin specific implementation.
Perhaps the way MyTourBook deals with Lap times is not the best one ?

What I understand from your code extract is that MyTouBook is calculating elapsed time and distance from time slice to next time slice : am I right ?
But when checking the Fit file data, we can see that there are specific fields to get directly the elapsed time/total time and the distance for each lap. Perhaps it would be better to use these data ?

![CaptureFitViewer](https://github.com/mytourbook/mytourbook/assets/64196685/ea0684be-532e-458d-8502-648a4783d299)

What do you think ?

**Comment by mff73** [comment]
i checked your file, and some of mine (suunto ones). I uploaded your file in runalyze to see, and laps are well managed.
In suunto fit files, a lap is starting at start_time and stopped at timestamp with a duration (total times (those ones may depend if there are pauses or not, but let simplify without pauses :) ).
MAYBE, garmin or strava or runalyze are not using timestamp to determine end of lap, but are just using start_time and lap duration.
Just a guess.

then, now, question : what may happen if there is a pause during a lap ? theorically, total_elapsed_time and total_timer_time will have to be used in coordination with pauses events.

**Comment by GurVJ** [comment]
> then, now, question : what may happen if there is a pause during a lap ? theorically, total_elapsed_time and total_timer_time will have to be used in coordination with pauses events.

It is the case in the last lap and the difference between total_timer_time and total_elapsed_time values tells us that there has been a "pause"  of 5:13.2 minutes.
![CaptureFitViewer2](https://github.com/mytourbook/mytourbook/assets/64196685/4a470e0b-30b3-46b4-b9b7-406d47ad4e24)

**Comment by FJBDev** [comment]
> What I understand from your code extract is that MyTouBook is calculating elapsed time and distance from time slice to next time slice : am I right ? But when checking the Fit file data, we can see that there are specific fields to get directly the elapsed time/total time and the distance for each lap. Perhaps it would be better to use these data ?

You can see [here](https://github.com/mytourbook/mytourbook/blob/364ae5d626fcf7ce1717b5e05c6835dafaba8527/bundles/net.tourbook.device.garmin.fit/src/net/tourbook/device/garmin/fit/listeners/MesgListener_Lap.java#L77) what MT uses to get the lap/marker time

**Reply by mff73** [reply]
i imported this fit file into Suunto ecosytem : and laps are well seen as well there, for duration and distance, but not for ascent/descent.
![image](https://github.com/mytourbook/mytourbook/assets/50132146/fb414db1-8179-4ebc-81e7-07fff957d573)

But, the definition of events is : 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/30f4cf20-b6b9-452a-838f-f95bbe834449)

so, i would more bet on a bug from this file : all timestamps are the same, and assuming that start_time is the only correct value.

**Comment by GurVJ** [comment]
> but not for ascent/descent.
Garmin Connect and Strava deals correctly with ascent/descent

> so, i would more bet on a bug from this file : all timestamps are the same, and assuming that start_time is the only correct value.
I hope you're right because currently I can't check my interval trainings anymore with MyToubook 😕
I am waiting impatiently for the next firmware official release.

**Comment by FJBDev** [comment]
> > so, i would more bet on a bug from this file : all timestamps are the same, and assuming that start_time is the only correct value.
> > I hope you're right because currently I can't check my interval trainings anymore with MyToubook 😕
> > I am waiting impatiently for the next firmware official release.

Since you say this above `After a quick test, it is also working with Strava`, what you can do for now is import a tour on Strava, export it and maybe you'll be able to see the laps correctly ?

I would be also willing to bet on the file being the issue since, AFAIK, the issue only happens for this firmware of this particular watch that only a few users have access to it.

> (I reported the problem but don't have any answer) 

Let us know what Garmin says

**Comment by FJBDev** [comment]
>(I reported the problem but don't have any answer)

@GurVJ 
Also, feel free to ask on the [official Garmin FIT SDK forum](https://forums.garmin.com/developer/fit-sdk/f/discussion), it seems that they respond to questions

**Comment by stvzldn** [comment]
This article by the5krunner might offer some insight:
[Garmin FIT Files – welcomed improvements for 2023/4. Is there a downside?](https://the5krunner.com/2024/02/12/garmin-fit-files-welcomed-improvements-for-2023-4-is-there-a-downside/)
Refers to changed storage of laps and other summary data.

**Reply by FJBDev** [reply]
@stvzldn Thanks for sharing , i was not aware of the "improvements".

**Reply by GurVJ** [reply]
@stvzldn , good catch.
Finally, when reading
"the timestamp of each message represents the time it was written to the file. (...) For files using the Summary First method, the timestamp approximates the file creation time. In either case, the timestamp should not be used to determine the start or end time of the summary message.",
it could say that the beta firmware I've got on my device is more a new implementation of FIT files than a bug...

**Comment by FJBDev** [comment]
> @stvzldn  "the timestamp of each message represents the time it was written to the file. (...) For files using the Summary First method, the timestamp approximates the file creation time. In either case, the timestamp should not be used to determine the start or end time of the summary message.", 

That explains it well. Now, I am wondering if that is a breaking change from Garmin or if there is still away to read the "previous" file messages with such a change

**Comment by mff73** [comment]
As far as i read it, each formula where timestamps are used should be revised and replaced by : 

```
To properly decode FIT Activity Files created using either the “Summary First” or “Summary Last” method, the End Time for the summary messages should be calculate as: ​

end_time = start_time + total_elasped_time ​
```
question : are all previous files reliable on the total_elpased_times ?
Am i correct ?

**Comment by GurVJ** [comment]
Just to make a quick  follow-up about FIT file implementation on my Garmin device : 
Garmin just released a new official firmware version of the Forerunner 965 (same firmware version for Forerunner 955 as well) and laps implementation of FIT files is the same as the last beta releases I tested.
So it was not a bug or an error but a real change in the way of implementing laps in FIT files.

**Comment by witdem** [comment]
I have now the same issue on my Garmin 265S 
The field starttime could be used, only it has the starttime instead of the endtime as the timestamp indicated. 
Starttime is available in older fit files as well (I checked some files from 2019)

**Comment by FJBDev** [comment]
> Same data from Garmin Connect software ![GarminConnectLaps](https://private-user-images.githubusercontent.com/64196685/301574612-632abc7d-1ac0-45ac-861a-05c95fddc9f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0ODk5NTQsIm5iZiI6MTcxMTQ4OTY1NCwicGF0aCI6Ii82NDE5NjY4NS8zMDE1NzQ2MTItNjMyYWJjN2QtMWFjMC00NWFjLTg2MWEtMDVjOTVmZGRjOWYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDIxNDczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1MTFjODgxYzRhMTZmOWM0ZGZlMWVkNGRhYWY0NmIwYTUyMTUyMWJlYTI5NjdlZTc2ZGMwYzlhYTU4NjA1OGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.X5unX2tmWVXm9Lfkd04d88p785S4m8TKJPgpyIOJOrw)
> 

@GurVJ  As I am taking a stab at this issue, it seems that with my fix the markers are getting imported with the new format but I still can't find why the first marker time is different than on your Garmin screenshot.
Notice it's different by 1 sec but maybe it's just a rounding difference from the platform you used to get this screenshot ? let me know since you originally created this file.

I could provide a test version to download if any of you are interested to test my fix

![image](https://github.com/mytourbook/mytourbook/assets/25235665/9312dd00-0157-40ef-b27c-72efd4fdb000)


Also, I should note that my fix is backward compatible, i.e.: It still imports fine the files that were generated with the “Summary Last” method (the unit tests verified that)

**Reply by mff73** [reply]
I can't test this particular fix (no garmin here), but if you do a testing version, I could eventually test Suunto battery values implemented by Wolfgang.

**Reply by FJBDev** [reply]
@mff73 Let's not "muddy the water" here please and mix up two different features.
If you want to test the Suunto battery values, you might want to ask in the appropriate discussion/issue [here](https://github.com/mytourbook/mytourbook/issues/1287)

**Comment by GurVJ** [comment]
> @GurVJ As I am taking a stab at this issue, it seems that with my fix the markers are getting imported with the new format but I still can't find why the first marker time is different than on your Garmin screenshot. Notice it's different by 1 sec but maybe it's just a rounding difference from the platform you used to get this screenshot ? let me know since you originally created this file.

@FJBDev, I spent some time on it but I think I finally found out why there are such differences between Fit data and Garmin Connnect data.
Actually Garmin Connect displays the "Split" messages data from FIT file and not the "Lap" messages data.
Split messages seem to decompose the activity into intervals with criterias that are not the lap trigger done on the device. Split types can be "interval_active", "interval_recovery", "interval_rest", "rwd_walk" (run/walk detection walking), etc. To me it is a "Garmin cooking" stuff.
That is why there is a difference between your fix and my screenshot (and your fix is what we need).

But I think there is still a problem of rounding in your fix as there is a difference between raw data of the FIT file and you screenshot.
To me the result of your fix should look like :

![image](https://github.com/mytourbook/mytourbook/assets/64196685/8360179e-2596-4874-8b7b-409210fe5f7f)

**Comment by FJBDev** [comment]
> @FJBDev, I spent some time on it but I think I finally found out why there are such differences between Fit data and Garmin Connnect data. Actually Garmin Connect displays the "Split" messages data from FIT file and not the "Lap" messages data. Split messages seem to decompose the activity into intervals with criterias that are not the lap trigger done on the device. Split types can be "interval_active", "interval_recovery", "interval_rest", "rwd_walk" (run/walk detection walking), etc. To me it is a "Garmin cooking" stuff. That is why there is a difference between your fix and my screenshot (and your fix is what we need).

Ok, that would explain the differences indeed
> 
> But I think there is still a problem of rounding in your fix as there is a difference between raw data of the FIT file and you screenshot. To me the result of your fix should look like :
> 
Possibly but I think the rounding difference is not related to the FIT file import but appears to be more a generalized display issue in MT that was also brought up [here](https://github.com/mytourbook/mytourbook/issues/820).
When I look in the code, it seems that internally we store the value in milliseconds but we only display them in seconds.

When I compare the raw data you put, it seems close minus the precision of milliseconds that it not displayed in the Marker View
Note that I used the TotalTimerTime to compare to the Marker View configured to display the recorded time

![image](https://github.com/mytourbook/mytourbook/assets/25235665/151ee605-3799-44af-a2ce-dabdd035576c)


![image](https://github.com/mytourbook/mytourbook/assets/25235665/2f734335-7e34-4138-8697-01e37ecb16e2)

![image](https://github.com/mytourbook/mytourbook/assets/25235665/d982f786-0c2d-4fde-8912-c20acdb7ed7f)

**Comment by GurVJ** [comment]
> Note that I used the TotalTimerTime to compare to the Marker View configured to display the recorded time

What I do not understand is when you use getTotalTimerTime() you get a result in seconds but it is a float (milliseconds given after the comma). So there must be a rounding to display it in second in your fix ?
Is it possible for you to log the result of the getTotalTimerTime() method for each lap ?

**Comment by FJBDev** [comment]
> What I do not understand is when you use getTotalTimerTime() you get a result in seconds but it is a float (milliseconds given after the comma). So there must be a rounding to display it in second in your fix ?

Correct, and this is what I was saying above, MT displays it in seconds even though it is stored in milliseconds when importing the FIT file

> Is it possible for you to log the result of the getTotalTimerTime() method for each lap ?

Below is the result for each lap

`2059.529
43.711
30.513
40.682
28.905
40.532
29.893
38.112
29.292
38.099
29.506
37.544
29.769
37.73
29.732
37.136
30.503
37.085
28.953
36.202
32.575
1225.877`

**Comment by GurVJ** [comment]
> Correct, and this is what I was saying above, MT displays it in seconds even though it is stored in milliseconds when importing the FIT file

OK but the way the result is displayed is not logical : sometimes the displayed value is the integer part of the getTotalTimerTime (43.711 => 0h43) and sometimes it is the rounded value (30.513 => 0h31) .
There is something weird here.

**Comment by FJBDev** [comment]
> OK but the way the result is displayed is not logical : sometimes the displayed value is the integer part of the getTotalTimerTime (43.711 => 0h43) and sometimes it is the rounded value (30.513 => 0h31) . There is something weird here.

Yes but this has nothing to do with the FIT Import. If you import a FIT with either the `Summary First` or `Summary Last`, you will see the same rounding being applied

**Comment by GurVJ** [comment]
You're right but it could be the opportunity to check where the problem comes from.
Just for my information (and if it is easy for you), would it be possible to extract from the database the values of the column TIME of the TOURMAKER Table for the tour you applied the fix to ?

**Comment by FJBDev** [comment]
> would it be possible to extract from the database the values of the column TIME of the TOURMAKER Table for the tour you applied the fix to ?

Sure, there you go

Note that per the code, this is how `time` is defined
```

/**
    * Time in seconds relative to the tour start. When value is not available it is set to
    * <code>-1</code>.
    */
```

"TIME"
2312
2355
2386
2426
2455
2496
2525
2564
2593
2631
2661
2698
2728
2765
2795
2833
2863
2901
2929
2966
2998
4536
